### PR TITLE
Add scan percentages (fixes #1032)

### LINKED ISSUE #1032: Show progress of scanning
During scanning Synching currently does not indicate a progress and stays this way for an indefinite amount of time. If possible, it would be great if "Scanning (XY %)" could be displayed (XY = scanned files / total files)

### PR DESCRIPTION
**Pull request by AudriusButkevicius**
Untested (need sleep), potentially bloats memory, but better than having to do database lookups twice (with a dry run), or bloating memory to cache data looked up in the database (lookup in dry run to decide if will hash, store somewhere, lookup from somewhere when checking if should hash).

### CODE REVIEW

**Comment by calmh** [comment]
```
08:20:21 <@calmh> [0]Jmr: i think we anyway buffer filenames in a bunch of places, so that should be ok
08:20:35 <@calmh> i saw the pull request but didn't have time to look at it yet
08:21:21 <@calmh> [0]Jmr: i don't know if there's some gain to use a shorter map key than the full filename, in case the filenames are long and 
                  many. i.e. using sha256(filename) as map key instead of the full thing
```

**Comment by AudriusButkevicius** [comment]
There is no map in this case, hence I am not sure what you are referring to.

**Comment by calmh** [comment]
Never mind, forget that, I'll look at what's actually implemented. My idea would require two scan passes anyway. :)

**Comment by AudriusButkevicius** [comment]
That doesn't work well as its double the db lookups.

File: lib/scanner/walk.go [hunk 1]
```diff
@@ -272,10 +347,10 @@ func (w *Walker) walkAndHashFiles(fchan, dchan chan protocol.FileInfo) filepath.
 			}
 
 			if debug {
-				l.Debugln("symlink to hash:", p, f)
+				l.Debugln("symlink changedb:", p, f)
```
Comment (calmh) [inline lib/scanner/walk.go]: "changed"?

File: lib/scanner/walk.go [hunk 2]
```diff
@@ -105,10 +113,77 @@ func (w *Walker) Walk() (chan protocol.FileInfo, error) {
...
+	ticker := time.NewTicker(time.Duration(w.ProgressTickIntervalS) * time.Second)
+
+	// We need to emit progres events, hence we create a routine which buffers
+	// the list of files to be hashed, counts the total number of
+	// bytes to hash, and once no more files need to be hashed (chan get closed),
+	// start a routine which periodically emits FolderScanProgress events,
+	// until a stop signal is sent by the parallel hasher.
+	// Parallel hasher is stopped by this routine when we close the channel over
+	// which it receives the files we ask it to hash.
+	go func() {
```
Comment (calmh) [inline lib/scanner/walk.go]: Do we need to provide updates more than once per file? If not, would it perhaps be cleaner to add more goroutines to the chain? E/goto:

FileInfos from the walker :arrow_right: Your buffering goroutine, keeps track of how much there is to scan :arrow_right: 
All the parallell hashers :arrow_right: Another goroutine that just intercepts the results from the hasher, bumps the progress counter according to that file's size and possibly generates the progress events :arrow_right: The usual database updater at the end of the chain...

The advantage of this is that we don't have to pass around a counter that needs atomic updates, we just need to calculate a total size and pass that to the interceptor goroutine when it starts...

Comment (AudriusButkevicius) [inline lib/scanner/walk.go]: That was my initial idea, but things like VM images will just stay dead for long periods of time. Unless we ramp it up a notch and report progress per file, which could also be cool.

File: gui/syncthing/core/syncthingController.js [hunk 3]
```diff
@@ -634,6 +650,14 @@ angular.module('syncthing.core')
             return Math.floor(pct);
         };
 
+        $scope.scanPercentage = function (folder) {
+            if (!$scope.scanProgress[folder]) {
+                return undefined;
+            }
+            var pct = 100 * %scope.scanProgress[folder].current / $scope.scanProgress[folder].total;
```
Comment (calmh) [inline gui/syncthing/core/syncthingController.js]: `$scope`

File: lib/scanner/walk.go [hunk 4]
```diff
@@ -105,10 +113,77 @@ func (w *Walker) Walk() (chan protocol.FileInfo, error) {
...
+	}
+
+	// Defaults to every 2 seconds.
+	if w.ProgressTickIntervalS == 0 {
+		w.ProgressTickIntervalS = 2
+	}
+
+	ticker := time.NewTicker(time.Duration(w.ProgressTickIntervalS) * time.Second)
+
+	// We need to emit progres events, hence we create a routine which buffers
```
Comment (calmh) [inline lib/scanner/walk.go]: "progress"

File: lib/scanner/walk.go [hunk 5]
```diff
@@ -105,10 +113,77 @@ func (w *Walker) Walk() (chan protocol.FileInfo, error) {
...
+	// the list of files to be hashed, counts the total number of
+	// bytes to hash, and once no more files need to be hashed (chan get closed),
```
Comment (calmh) [inline lib/scanner/walk.go]: "gets"

**Comment by calmh** [comment]
Apart from the super anal retentive comment typo comments and the one thing in the javascript, LGTM :+1:

File: lib/scanner/walk.go (hunk 2 above)
Comment (calmh) [inline lib/scanner/walk.go]: :ok_hand:

**Comment by AudriusButkevicius** [comment]
Fixed. Seems to do what it says, feel free to merge if you are happy.

**Comment by calmh** [comment]
Good stuff :cookie:

**Pull request merged**